nengo_loihi/splitter.py

```diff
@@ -4,184 +4,280 @@
-            return PassthroughSplit(network, ignore)
-        else:
-            return PassthroughSplit(None)
+    def __init__(self, network, hostchip, passthrough, strict):
+        self.hostchip = hostchip
+        self.passthrough = passthrough
 
-    def _place_objects(self, network):
-        # Step 1. Place nodes on host
-        self._seen_objects.update(network.all_nodes)
+        self.objs = set()
+        self._precomputable = True
+        self._conns = (
+            set(network.all_connections) | self.passthrough.to_add
+        ) - self.passthrough.to_remove
 
-        # Step 2. Place all possible ensembles on chip
-        # Note: assumes add_params already called by the simulator
-        for ens in network.all_ensembles:
-            if network.config[ens].on_chip in (None, True) and not isinstance(
-                ens.neuron_type, Direct
-            ):
-                self._chip_objects.add(ens)
-            self._seen_objects.add(ens)
+        # Learning rules are not supported with precompute=True because
+        # this would require a hybrid simulation where some parts of the
+        # model interact with the host while other parts are precomputed
+        # ahead of time. The simulator assumes that precompute=True does not
+        # require any interaction between host and chip between time-steps.
+        # Also see issue #214.
+        has_learning = any(conn.learning_rule is not None for conn in self._conns)
 
-        # Step 3. Move learning ensembles (post and error) to host
-        for conn in network.all_connections:
-            pre = base_obj(conn.pre)
-            post = base_obj(conn.post)
-            if (
-                conn.learning_rule_type is not None
-                and isinstance(post, Ensemble)
-                and post in self._chip_objects
-            ):
-                if network.config[post].on_chip:
-                    raise BuildError(
-                        "Post ensemble (%r) of learned "
-                        "connection (%r) must not be configured "
-                        "as on_chip." % (post, conn)
-                    )
-                self._chip_objects.remove(post)
-            elif (
-                isinstance(post, LearningRule)
-                and isinstance(pre, Ensemble)
-                and pre in self._chip_objects
-            ):
-                if network.config[pre].on_chip:
-                    raise BuildError(
-                        "Pre ensemble (%r) of error "
-                        "connection (%r) must not be configured "
-                        "as on_chip." % (pre, conn)
-                    )
-                self._chip_objects.remove(pre)
+        # On-chip convolutional connections are also not supported with
+        # precompute=True because the BasicSpikeGenerator cannot send population
+        # spikes correctly, meaning we have to use SNIPs.
+        # TODO: check with Eric if this logic is correct.
```

Comment:
First of all, it's only host->chip connections that are the problem. Connections that are fully on the chip are fine. So if someone sets `precompute=True` on a network that only has convolution connections that are fully on the chip, will this be a problem?

Second, we do add support for BasicSpikeGenerator sending spikes to the chip in #261, so this logic will have to change once we get there.

Reply:
Yes, the chip->chip connections will fail with this. The reason I thought that that would be OK is because all of the tests that we were doing using convolutions explicitly set `precompute=False`, so presumably anything with convolutions isn't precomputable. But if that's not the case then the comment is incorrect and we should only be checking for host->chip convolutional connections, which is easily changed.

Should be trivial to make changes to this logic when things change, don't think this addition makes a difference.

Reply:
Done in e77ece72